kernel/aster-nix/src/process/sync/condvar.rs

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::VecDeque, sync::Arc};
+use core::{
+    sync::atomic::{AtomicBool, Ordering},
+    time::Duration,
+};
+
+use aster_frame::sync::{guard_lock, MutexGuard, SpinLock};
+
+use super::Pauser;
+use crate::Errno;
+
+/// LockResult, different from Rust std.
+/// The result of a lock operation.
+pub enum LockResult<Guard> {
+    Ok(Guard),
+    Timeout(Guard),
+    Interrupted(Guard),
+    Unknown(Guard),
+}
+
+impl<Guard> LockResult<Guard> {
+    pub fn unwrap(self) -> Guard {
+        match self {
+            LockResult::Ok(guard) => guard,
+            LockResult::Timeout(guard) => guard,
+            LockResult::Interrupted(guard) => guard,
+            LockResult::Unknown(guard) => guard,
+        }
+    }
+}
+
+/// A condition variable.
+pub struct Condvar {
+    pauser: Arc<Pauser>,
+    should_wait: SpinLock<VecDeque<Arc<AtomicBool>>>,
```

Comment:
That's a good idea. I initially wanted to use `BitSet` to implement this. This would be much more light-weight, but it would also limit the number of threads, so I chose `VecDeque`.

Reply:
I am not convinced, yet. I am keeping this issue open.